opm/simulators/wells/ParallelWellInfo.hpp

```diff
@@ -0,0 +1,260 @@
+{
+
+/// \brief Class to facilitate getting values associated with the above perforation
+///
+class CommunicateAbove
+{
+    enum Attribute {
+      owner=1, overlap=2
+    };
+    using MPIComm = typename Dune::MPIHelper::MPICommunicator;
+#if DUNE_VERSION_NEWER(DUNE_COMMON, 2, 7)
+    using Communication = Dune::Communication<MPIComm>;
+#else
+    using Communication = Dune::CollectiveCommunication<MPIComm>;
+#endif
+    using LocalIndex = Dune::ParallelLocalIndex<Attribute>;
+#if HAVE_MPI
+    using IndexSet = Dune::ParallelIndexSet<std::size_t,LocalIndex,50>;
+    using RI = Dune::RemoteIndices<IndexSet>;
+#endif
+
+public:
+    CommunicateAbove(const Communication& comm);
+    /// \brief Adds information about the ecl indices of the perforations.
+    ///
+    /// \warning Theses indices need to be push in the same order as they
+    ///          appear in the ECL well specifiation. Use -1 if there is
+    ///          no perforation above.
+    /// \param above The ECL index of the next open perforation above.
+    /// \param current The ECL index of the current open perforation.
+    void pushBackEclIndex(int above, int current, bool owner=true);
+
+    /// \brief Clear all the parallel information
+    void clear();
+
+    /// \brief Indicates that we will add the index information
+    /// \see pushBackEclIndex
+    void beginReset();
+
+    /// \brief Indicates that the index information is complete.
+    ///
+    /// Sets up the commmunication structures to be used by
+    /// communicate()
+    void endReset();
+
+    /// \brief Creates an array of values for the perforation above.
+    /// \param first_value Value to use for above of the first perforation
+    /// \param current C-array of the values at the perforations
+    /// \param size The size of the C-array and the returned vector
+    /// \return a vector containing the values for the perforation above.
+    std::vector<double> communicate(double first_value,
+                                    const double* current,
+                                    std::size_t size);
+
+private:
+#if HAVE_MPI
+    Communication comm_;
+    /// \brief Mapping of the local well index to ecl index
+    IndexSet current_indices_;
+    /// \brief Mapping of the above well index to ecl index
+    IndexSet above_indices_;
+    RI remote_indices_;
+    Dune::Interface interface_;
+    Dune::BufferedCommunicator communicator_;
+#endif
+    std::size_t count_{};
+};
+
+
+/// \brief Class encapsulating some information about parallel wells
+///
+/// e.g. It provides a communicator for well information
+class ParallelWellInfo
+{
+public:
+    using MPIComm = typename Dune::MPIHelper::MPICommunicator;
+#if DUNE_VERSION_NEWER(DUNE_COMMON, 2, 7)
+    using Communication = Dune::Communication<MPIComm>;
+#else
+    using Communication = Dune::CollectiveCommunication<MPIComm>;
+#endif
+
+
+    /// \brief Constructs object using MPI_COMM_SELF
+    ParallelWellInfo(const std::string& name = {""},
```

Comment:
It is needed.

Reply:
> It is needed.

Why? Certainly you will not construct this in 5 different ways? Why the complexity?

Reply:
Thanks for the usage notes. I do not see the need for the default constructor still, but I assume it will be needed.

> Seriously, how is using default arguments in two constructors increasing complexity. In my opinion it is the other way around.

Because it adds more possible signatures. More signatures means more complexity reading and understanding the code. I may not have complained about too many constructors much, but then it is not a very common problem!

I think default arguments should preferably only be used when adding a new argument and it is important to not have to rewrite existing calls.